sdks/java/io/contextualtextio/src/main/java/org/apache/beam/sdk/io/contextualtextio/ContextualTextIO.java

```diff
@@ -423,90 +422,91 @@ public void processElement(
-          if (a.getKey().compareTo(b.getKey()) == 0) {
-            return a.getValue().compareTo(b.getValue());
-          }
-          return a.getKey().compareTo(b.getKey());
-        }
-      }
-
       @ProcessElement
       public void processElement(ProcessContext p) {
-        // Process each file from which is a key from the side input
-
-        // Get the Map Containing the size from side-input
+        // Get the multimap side input containing the size of each read range.
         Map<String, Iterable<KV<Long, Long>>> rangeSizesMap = p.sideInput(rangeSizes);
 
+        // Process each file, retrieving each filename as key from the side input.
         for (Entry<String, Iterable<KV<Long, Long>>> entrySet : rangeSizesMap.entrySet()) {
-          // The FileRange Pair must be sorted
-          SortedMap<KV<String, Long>, Long> sorted = new TreeMap<>(new FileRangeComparator<>());
+          // The offsets in the file must be sorted.
+          SortedMap<Long, Long> sorted = new TreeMap<>();
 
           entrySet
               .getValue()
               .iterator()
-              .forEachRemaining(
-                  x -> sorted.put(KV.of(entrySet.getKey(), x.getKey()), x.getValue()));
-
-          // HashMap that tracks number of records passed for each file
-          Map<String, Long> pastRecords = new HashMap<>();
-
-          // For each (File, Range) Pair, compute the number of records before it
-          for (Map.Entry<KV<String, Long>, Long> entry : sorted.entrySet()) {
-            Long numRecords = entry.getValue();
-            KV<String, Long> fileRange = entry.getKey();
-            String file = fileRange.getKey();
-            Long numRecordsBefore = 0L;
-            if (pastRecords.containsKey(file)) {
-              numRecordsBefore = pastRecords.get(file);
-            }
-            p.output(KV.of(fileRange, numRecordsBefore));
-            pastRecords.put(file, numRecordsBefore + numRecords);
+              .forEachRemaining(x -> sorted.put(x.getKey(), x.getValue()));
+
+          String file = entrySet.getKey();
+          Map<Long, Long> numRecordsBeforeEachOffset = new HashMap<>();
+          Long numRecordsBefore = 0L;
+          for (Map.Entry<Long, Long> entry : sorted.entrySet()) {
+            Long offset = entry.getKey();
+            Long numRecordsInRangeStartingAtThisOffset = entry.getValue();
+            numRecordsBeforeEachOffset.put(offset, numRecordsBefore);
+            numRecordsBefore += numRecordsInRangeStartingAtThisOffset;
           }
+          p.output(KV.of(file, numRecordsBeforeEachOffset));
         }
       }
     }
 
+    /**
+     * Helper transform for computing absolute position of each record given the read range of each
+     * record and a side input describing the describing number of records that precede the
+     * beginning of each read range.
+     */
     static class AssignRecordNums extends DoFn<KV<KV<String, Long>, Row>, Row> {
```

Comment:
Can we leave a TODO for a performance test to be added as part of the performance suite?